emacs/.config/emacs/config.org
#+Title: Satori's GNU Emacs Config
#+AUTHOR: Satori
#+DESCRIPTION: Satori's personal Emacs config.
#+STARTUP: content
#+OPTIONS: TOC:2

* References
** [[https://github.com/daviwil/emacs-from-scratch/][Emacs From Scratch]]
** [[https://gitlab.com/dwt1/configuring-emacs][Configuring Emacs]]
* Configuration

** Source Scripts
#+begin_src emacs-lisp
(add-to-list 'load-path "~/.config/emacs/scripts/")
(require 'elpaca-setup)
(require 'buffer-move)
#+end_src
** Initial Sets
#+begin_src emacs-lisp
(setq vc-follow-symlinks t)
#+end_src
** Scratch Buffer
#+begin_src emacs-lisp
(setq initial-scratch-message "")
#+end_src
** Graphics
*** Font
**** Default
#+begin_src emacs-lisp
(set-face-attribute 'default nil
		    :font "Fira Code"
		    :height 120
		    :weight 'medium)
#+end_src
*** Theme
**** Catppuccin
#+begin_src emacs-lisp
(use-package catppuccin-theme :after org :config
  (load-theme 'catppuccin :no-confirm))
#+end_src
*** Rainbow Mode
#+begin_src emacs-lisp
(use-package rainbow-mode
  :diminish
  :hook org-mode prog-mode)
#+end_src

*** Disable Menubar, Toolbars and Scrollbars
#+begin_src emacs-lisp
(menu-bar-mode -1)
(tool-bar-mode -1)
(scroll-bar-mode -1)
#+end_src
*** Display Line Numbers and Truncated Lines
#+begin_src emacs-lisp
(global-display-line-numbers-mode 1)
(global-visual-line-mode t)
#+end_src
*** Visual Bell
#+begin_src emacs-lisp
(setq visible-bell t)
#+end_src
*** All The Icons
#+begin_src emacs-lisp
(use-package all-the-icons
  :ensure t
  :if (display-graphic-p))

(use-package all-the-icons-dired
  :hook (dired-mode . (lambda () (all-the-icons-dired-mode t))))
#+end_src
*** Modeline
#+begin_src emacs-lisp
(use-package doom-modeline
  :ensure t
  :init (doom-modeline-mode 1)
  :config
  (setq doom-modeline-height 10      ;; sets modeline height
        doom-modeline-bar-width 0    ;; sets right bar width
        doom-modeline-persp-name t   ;; adds perspective name to modeline
        doom-modeline-persp-icon 'nil)) ;; adds folder icon next to persp name
#+end_src
** Org Mode
*** Configuration
#+begin_src emacs-lisp
(setq org-agenda-files
      '("~/Documents/notes/"))
(setq org-agenda-block-separator 8411)
(require 'org)
(setq org-display-custom-times t)
(setq org-time-stamp-custom-formats '("<%a %b %e %Y>" . "<%a %e %b %Y %I:%M %p"))
(add-hook 'org-mode-hook 'org-indent-mode)

(defun update-last-modified-date ()
  "Update the #+LAST_MODIFIED: property in org files."
  (when (and (eq major-mode 'org-mode)
             (buffer-file-name))
    (save-excursion
      (goto-char (point-min))
      (when (re-search-forward "^#\\+last_modified:" nil t)
        (kill-line)
        (insert (format-time-string " [%Y-%m-%d %a %H:%M]"))))))

(add-hook 'before-save-hook 'update-last-modified-date)

#+end_src
*** Org Roam
#+begin_src emacs-lisp

(use-package org-roam
  :ensure t
  :custom
  (org-roam-directory "~/Documents/org-roam")
  (org-roam-dailies-directory "~/Documents/org-roam/daily")
  :config
  (setq org-roam-node-display-template
	(concat "${title:*} "
		(propertize "${tags:10}" 'face 'org-tag)))

  (setq org-roam-capture-templates
	'(("d" "default" plain
	   "%?"
	   :if-new (file+head "%<%Y%m%d%H%M%S>-${slug}.org" "#+title: ${title}\n#+created:%u\n#+last_motified:%U\n")
	   :unnarrowed t)))

  (org-roam-db-autosync-mode)
  (org-roam-setup))

#+end_src
*** Enable Org Bullets
#+begin_src emacs-lisp
(use-package org-bullets
  :hook (org-mode . org-bullets-mode)
  :custom
  (org-bullets-bullet-list '("◉" "○" "●" "○" "●" "○" "●")))
#+end_src
*** Disable Electric Indent
#+begin_src emacs-lisp
(electric-indent-mode -1)
(setq org-edit-src-content-indentation 0)
#+end_src
*** Enable Table of Contents
#+begin_src emacs-lisp
(use-package toc-org
  :commands toc-org-enable
  :init (add-hook 'org-mode-hook 'toc-org-enable))
#+end_src
*** Source Code Block Tag Expansion

| Typing the below + TAB | Expands to ...                          |
|------------------------+-----------------------------------------|
| <a                     | '#+BEGIN_EXPORT ascii' … '#+END_EXPORT  |
| <c                     | '#+BEGIN_CENTER' … '#+END_CENTER'       |
| <C                     | '#+BEGIN_COMMENT' … '#+END_COMMENT'     |
| <e                     | '#+BEGIN_EXAMPLE' … '#+END_EXAMPLE'     |
| <E                     | '#+BEGIN_EXPORT' … '#+END_EXPORT'       |
| <h                     | '#+BEGIN_EXPORT html' … '#+END_EXPORT'  |
| <l                     | '#+BEGIN_EXPORT latex' … '#+END_EXPORT' |
| <q                     | '#+BEGIN_QUOTE' … '#+END_QUOTE'         |
| <s                     | '#+BEGIN_SRC' … '#+END_SRC'             |
| <v                     | '#+BEGIN_VERSE' … '#+END_VERSE'         |

#+begin_src emacs-lisp
(require 'org-tempo)
#+end_src
*** Auto Tangle
#+begin_src emacs-lisp
(use-package org-auto-tangle
  :defer t
  :hook (org-mode . org-auto-tangle-mode)
  :config (setq org-auto-tangle-default t))
#+end_src
** Load Evil Mode
#+begin_src emacs-lisp
(use-package evil
  :init
  (setq evil-want-integration t
        evil-want-keybinding nil
        evil-vsplit-window-right t
        evil-split-window-below t
        evil-undo-system 'undo-redo)
  (evil-mode))

(use-package evil-collection
  :after evil
  :config
  (add-to-list 'evil-collection-mode-list 'help)
  (evil-collection-init))

(use-package evil-tutor)
#+end_src
** KeyBinds
*** Mini Buffer Escape
#+begin_src emacs-lisp
(global-set-key [escape] 'keyboard-escape-quit)
#+end_src
*** Zoom In/Out
#+begin_src emacs-lisp
(global-set-key (kbd "C-=") 'text-scale-increase)
(global-set-key (kbd "C--") 'text-scale-decrease)
#+end_src
*** General Keybindings
#+begin_src emacs-lisp
(use-package general
  :config
  (general-evil-setup)
  ;; set up 'SPC' as the global leader key
  (general-create-definer satori/leader-keys
    :states '(normal insert visual emacs)
    :keymaps 'override
    :prefix "SPC" ;; set leader
    :global-prefix "M-SPC") ;; access leader in insert mode

  (satori/leader-keys
    "SPC" '(counsel-M-x :wk "Counsel M-x")
    "ff" '(find-file :wk "Find file")
    "fc" '((lambda () (interactive) (find-file "~/.config/emacs/config.org")) :wk "Edit emacs config")
    "fr" '(counsel-recentf :wk "Find recent files")
    "TAB TAB" '(comment-line :wk "Comment lines")
    "s" '(save-buffer :wk "Save file")
    "d" '(dashboard-open :wk "Open Dashboard"))

  (satori/leader-keys
    "g" '(:ignore t :wk "Git")
    "g/" '(magit-displatch :wk "Magit dispatch")
    "g." '(magit-file-displatch :wk "Magit file dispatch")
    "gb" '(magit-branch-checkout :wk "Switch branch")
    "gc" '(:ignore t :wk "Create")
    "gcb" '(magit-branch-and-checkout :wk "Create branch and checkout")
    "gcc" '(magit-commit-create :wk "Create commit")
    "gcf" '(magit-commit-fixup :wk "Create fixup commit")
    "gC" '(magit-clone :wk "Clone repo")
    "gf" '(:ignore t :wk "Find")
    "gfc" '(magit-show-commit :wk "Show commit")
    "gff" '(magit-find-file :wk "Magit find file")
    "gfg" '(magit-find-git-config-file :wk "Find gitconfig file")
    "gF" '(magit-fetch :wk "Git fetch")
    "gg" '(magit-status :wk "Magit status")
    "gi" '(magit-init :wk "Initialize git repo")
    "gl" '(magit-log-buffer-file :wk "Magit buffer log")
    "gr" '(vc-revert :wk "Git revert file")
    "gs" '(magit-stage-file :wk "Git stage file")
    "gt" '(git-timemachine :wk "Git time machine")
    "gu" '(magit-stage-file :wk "Git unstage file"))

  (satori/leader-keys
    "b" '(:ignore t :wk "buffer")
    "bb" '(switch-to-buffer :wk "Switch buffer")
    "bi" '(ibuffer :wk "Ibuffer")
    "bk" '(kill-this-buffer :wk "Kill this buffer")
    "bn" '(next-buffer :wk "Next buffer")
    "bp" '(previous-buffer :wk "Previous buffer")
    "br" '(revert-buffer :wk "Reload buffer"))

  (satori/leader-keys
    "k" '(:ignore t :wk "Bookmarks")
    "ks" '(bookmark-set :wk "Set Bookmark")
    "kj" '(bookmark-jump :wk "Jump Bookmark")
    "kl" '(list-bookmarks :wk "List Bookmark")
    "kd" '(bookmark-delete :wk "Delete Bookmarks"))

  (satori/leader-keys
    "e" '(:ignore t :wk "Eshell/Evaluate")
    "eb" '(eval-buffer :wk "Evaluate elisp in buffer")
    "ed" '(eval-defun :wk "Evaluate defun containing or after point")
    "ee" '(eval-expression :wk "Evaluate and elisp expression")
    "eh" '(counsel-esh-history :which-key "Eshell History")
    "el" '(eval-last-sexp :wk "Evaluate elisp expression before point")
    "er" '(eval-region :wk "Evaluate elisp in region"))

  (satori/leader-keys
    "h" '(:ignore t :wk "Help")
    "hd" '(:ignore t :wk "Describe")
    "hdf" '(describe-function :wk "Describe function")
    "hdv" '(describe-variable :wk "Describe variable")
    "hdk" '(describe-key :wk "Describe key")
    "hdb" '(describe-key-briefly :wk "Describe key briefly")
    "hi" '(:ignore t :wk "Info")
    "hii" '(info :wk "Info documentation browser")
    "hid" '(Info-directory :wk "Info Directory")
    "hia" '(info-apropos :wk "Find indice")
    "hr" '(+reload-init :wk "Reload emacs config")
    "ht" '(evil-tutor-start :wk "Evil tutor"))

  (satori/leader-keys
    "o" '(:ignore t :wk "Org")
    "oa" '(org-agenda :wk "Org agenda")
    "os" '((lambda () (interactive) (find-file "~/Documents/org-roam/20231110004145-satori.org")) :wk "Satori's Brain")
    "oe" '(org-export-dispatch :wk "Org export dispatch")
    "oi" '(org-toggle-item :wk "Org toggle item")
    "ot" '(org-todo :wk "Org todo")
    "oB" '(org-babel-tangle :wk "Org babel tangle")
    "oT" '(org-todo-list :wk "Org todo list")
    "ol" '(org-insert-link :wk "Org insert link"))

  (satori/leader-keys
    "r" '(:ignore t :wk "Roam")
    "rf" '(org-roam-node-find :wk "Roam find")
    "ri" '(org-roam-node-insert :wk "Roam insert"))

  (satori/leader-keys
    "ob" '(:ignore t :wk "Tables")
    "ob -" '(org-table-insert-hline :wk "Insert hline in table"))

  (satori/leader-keys
    "od" '(:ignore t :wk "Date/deadline")
    "odt" '(org-time-stamp :wk "Org time stamp"))

  (satori/leader-keys
    "l" '(:ignore t :wk "Misc Temp")
    "ls" '(lsp-treemacs-symbols :wk "Toggle treemacs-symbols"))

  (satori/leader-keys
    "t" '(:ignore t :wk "Toggle")
    "tl" '(display-line-numbers-mode :wk "Toggle line numbers")
    "tt" '(visual-line-mode :wk "Toggle truncated lines")
    "t/" '(vterm-toggle :wk "Toggle vterm")
    "te" '(eshell-toggle :wk "Toggle eshell")
    "tm" '(treemacs :wk "Toggle treemacs"))

  (satori/leader-keys
    "w" '(:ignore t :wk "Windows")
    ;; Window splits
    "wd" '(evil-window-delete :wk "Close window")
    "wa" '(evil-window-new :wk "New window")
    "wv" '(evil-window-vsplit :wk "Vertical split window")
    "ws" '(evil-window-split :wk "Horizontal split window")
    ;; Window motions
    "ww" '(evil-window-next :wk "Goto next window")
    ;; Move Windows
    "wH" '(buf-move-left :wk "Buffer move left")
    "wJ" '(buf-move-down :wk "Buffer move down")
    "wK" '(buf-move-up :wk "Buffer move up")
    "wL" '(buf-move-right :wk "Buffer move right"))
  (general-define-key
   :keymaps 'global
   "C-l" 'evil-window-right
   "C-h" 'evil-window-left
   "C-j" 'evil-window-up
   "C-k" 'evil-window-down)
  )
#+End_src

** Git
*** Time Machine
#+begin_src emacs-lisp
(use-package git-timemachine
  :after git-timemachine
  :hook (evil-normalize-keymaps . git-timemachine-hook)
  :config
  (evil-define-key 'normal git-timemachine-mode-map (kbd "C-j") 'git-timemachine-show-previous-revision)
  (evil-define-key 'normal git-timemachine-mode-map (kbd "C-k") 'git-timemachine-show-next-revision))
#+end_src
*** Magit
[[https://magit.vc/manual/][Magit]] is a full-featured git client.
#+begin_src emacs-lisp
(use-package magit)
#+end_src
** Shells & Terminals
*** Eshell
#+begin_src emacs-lisp
(add-hook 'eshell-mode-hook (lambda ()
                              (setq display-line-numbers-type nil)
                              (setq mode-line-format nil)))

(use-package eshell-toggle
  :custom
  (eshell-toggle-size-fraction 3)
  (eshell-toggle-use-projectile-root t)
  (eshell-toggle-run-command nil)
  (eshell-toggle-init-function #'eshell-toggle-init-eshell))

(use-package eshell-syntax-highlighting
  :after esh-mode
  :config
  (eshell-syntax-highlighting-global-mode +1))

;; eshell-syntax-highlighting -- adds fish/zsh-like syntax highlighting.
;; eshell-rc-script -- your profile for eshell; like a bashrc for eshell.
;; eshell-aliases-file -- sets an aliases file for the eshell.

(setq eshell-rc-script (concat user-emacs-directory "eshell/profile")
      eshell-aliases-file (concat user-emacs-directory "eshell/aliases")
      eshell-history-size 5000
      eshell-buffer-maximum-lines 5000
      eshell-hist-ignoredups t
      eshell-scroll-to-bottom-on-input t
      eshell-destroy-buffer-when-process-dies t
      eshell-visual-commands'("bash" "fish" "htop" "ssh" "top" "zsh"))
#+end_src
*** Vterm
#+begin_src emacs-lisp
(defun disable-line-numbers ()
  (display-line-numbers-mode -1))

(defun disable-mode-line ()
  (setq mode-line-format nil))

(use-package vterm
  :hook (vterm-mode . disable-line-numbers)
  :hook (vterm-mode . disable-mode-line)
  :config
  (setq shell-file-name "/bin/zsh"
	vterm-max-scrollback 5000))
#+end_src
*** Vterm-Toggle
#+begin_src emacs-lisp
(use-package vterm-toggle
  :after vterm
  :config
  (setq vterm-toggle-fullscreen-p nil)
  (setq vterm-toggle-scope 'project)
  (add-to-list 'display-buffer-alist
	       '((lambda (buffer-or-name _)
		   (let ((buffer (get-buffer buffer-or-name)))
		     (with-current-buffer buffer
		       (or (equal major-mode 'vterm-mode)
			   (string-prefix-p vterm-buffer-name (buffer-name buffer))))))
		 (display-buffer-reuse-window display-buffer-at-bottom)
		 (reusable-frames . visible)
		 (window-height . 0.3))))
#+end_src

emacs/.config/emacs/config.el
(add-to-list 'load-path "~/.config/emacs/scripts/")
(require 'elpaca-setup)
(require 'buffer-move)

(setq vc-follow-symlinks t)

(setq initial-scratch-message "")

(set-face-attribute 'default nil
		    :font "Fira Code"
		    :height 120
		    :weight 'medium)

(use-package catppuccin-theme :after org :config
  (load-theme 'catppuccin :no-confirm))

(use-package rainbow-mode
  :diminish
  :hook org-mode prog-mode)

(menu-bar-mode -1)
(tool-bar-mode -1)
(scroll-bar-mode -1)

(global-display-line-numbers-mode 1)
(global-visual-line-mode t)

(setq visible-bell t)

(use-package all-the-icons
  :ensure t
  :if (display-graphic-p))

(use-package all-the-icons-dired
  :hook (dired-mode . (lambda () (all-the-icons-dired-mode t))))

(use-package doom-modeline
  :ensure t
  :init (doom-modeline-mode 1)
  :config
  (setq doom-modeline-height 10      ;; sets modeline height
        doom-modeline-bar-width 0    ;; sets right bar width
        doom-modeline-persp-name t   ;; adds perspective name to modeline
        doom-modeline-persp-icon 'nil)) ;; adds folder icon next to persp name

(setq org-agenda-files
      '("~/Documents/notes/"))
(setq org-agenda-block-separator 8411)
(require 'org)
(setq org-display-custom-times t)
(setq org-time-stamp-custom-formats '("<%a %b %e %Y>" . "<%a %e %b %Y %I:%M %p"))
(add-hook 'org-mode-hook 'org-indent-mode)

(defun update-last-modified-date ()
  "Update the #+LAST_MODIFIED: property in org files."
  (when (and (eq major-mode 'org-mode)
             (buffer-file-name))
    (save-excursion
      (goto-char (point-min))
      (when (re-search-forward "^#\\+last_modified:" nil t)
        (kill-line)
        (insert (format-time-string " [%Y-%m-%d %a %H:%M]"))))))

(add-hook 'before-save-hook 'update-last-modified-date)

(use-package org-roam
:ensure t
:custom
(org-roam-directory "~/Documents/org-roam")
(org-roam-dailies-directory "~/Documents/org-roam/daily")
:config
(setq org-roam-node-display-template
      (concat "${title:*} "
              (propertize "${tags:10}" 'face 'org-tag)))

(setq org-roam-capture-templates
    '(("d" "default" plain
       "%?"
       :if-new (file+head "%<%Y%m%d%H%M%S>-${slug}.org" "#+title: ${title}\n#+created:%u\n#+last_motified:%U\n")
       :unnarrowed t)))

(org-roam-db-autosync-mode)
(org-roam-setup))

(use-package org-bullets
  :hook (org-mode . org-bullets-mode)
  :custom
  (org-bullets-bullet-list '("◉" "○" "●" "○" "●" "○" "●")))

(electric-indent-mode -1)
(setq org-edit-src-content-indentation 0)

(use-package toc-org
  :commands toc-org-enable
  :init (add-hook 'org-mode-hook 'toc-org-enable))

(require 'org-tempo)

(use-package org-auto-tangle
  :defer t
  :hook (org-mode . org-auto-tangle-mode)
  :config (setq org-auto-tangle-default t))

(use-package evil
    :init
    (setq evil-want-integration t
          evil-want-keybinding nil
          evil-vsplit-window-right t
          evil-split-window-below t
          evil-undo-system 'undo-redo)
    (evil-mode))

(use-package evil-collection
  :after evil
  :config
  (add-to-list 'evil-collection-mode-list 'help)
  (evil-collection-init))

(use-package evil-tutor)

(global-set-key [escape] 'keyboard-escape-quit)

(global-set-key (kbd "C-=") 'text-scale-increase)
(global-set-key (kbd "C--") 'text-scale-decrease)

(use-package general
  :config
  (general-evil-setup)
  ;; set up 'SPC' as the global leader key
  (general-create-definer satori/leader-keys
    :states '(normal insert visual emacs)
    :keymaps 'override
    :prefix "SPC" ;; set leader
    :global-prefix "M-SPC") ;; access leader in insert mode

  (satori/leader-keys
    "SPC" '(counsel-M-x :wk "Counsel M-x")
    "ff" '(find-file :wk "Find file")
    "fc" '((lambda () (interactive) (find-file "~/.config/emacs/config.org")) :wk "Edit emacs config")
    "fr" '(counsel-recentf :wk "Find recent files")
    "TAB TAB" '(comment-line :wk "Comment lines")
    "s" '(save-buffer :wk "Save file")
    "d" '(dashboard-open :wk "Open Dashboard"))

  (satori/leader-keys
    "g" '(:ignore t :wk "Git")
    "g/" '(magit-displatch :wk "Magit dispatch")
    "g." '(magit-file-displatch :wk "Magit file dispatch")
    "gb" '(magit-branch-checkout :wk "Switch branch")
    "gc" '(:ignore t :wk "Create")
    "gcb" '(magit-branch-and-checkout :wk "Create branch and checkout")
    "gcc" '(magit-commit-create :wk "Create commit")
    "gcf" '(magit-commit-fixup :wk "Create fixup commit")
    "gC" '(magit-clone :wk "Clone repo")
    "gf" '(:ignore t :wk "Find")
    "gfc" '(magit-show-commit :wk "Show commit")
    "gff" '(magit-find-file :wk "Magit find file")
    "gfg" '(magit-find-git-config-file :wk "Find gitconfig file")
    "gF" '(magit-fetch :wk "Git fetch")
    "gg" '(magit-status :wk "Magit status")
    "gi" '(magit-init :wk "Initialize git repo")
    "gl" '(magit-log-buffer-file :wk "Magit buffer log")
    "gr" '(vc-revert :wk "Git revert file")
    "gs" '(magit-stage-file :wk "Git stage file")
    "gt" '(git-timemachine :wk "Git time machine")
    "gu" '(magit-stage-file :wk "Git unstage file"))

  (satori/leader-keys
    "b" '(:ignore t :wk "buffer")
    "bb" '(switch-to-buffer :wk "Switch buffer")
    "bi" '(ibuffer :wk "Ibuffer")
    "bk" '(kill-this-buffer :wk "Kill this buffer")
    "bn" '(next-buffer :wk "Next buffer")
    "bp" '(previous-buffer :wk "Previous buffer")
    "br" '(revert-buffer :wk "Reload buffer"))

  (satori/leader-keys
    "k" '(:ignore t :wk "Bookmarks")
    "ks" '(bookmark-set :wk "Set Bookmark")
    "kj" '(bookmark-jump :wk "Jump Bookmark")
    "kl" '(list-bookmarks :wk "List Bookmark")
    "kd" '(bookmark-delete :wk "Delete Bookmarks"))

  (satori/leader-keys
    "e" '(:ignore t :wk "Eshell/Evaluate")
    "eb" '(eval-buffer :wk "Evaluate elisp in buffer")
    "ed" '(eval-defun :wk "Evaluate defun containing or after point")
    "ee" '(eval-expression :wk "Evaluate and elisp expression")
    "eh" '(counsel-esh-history :which-key "Eshell History")
    "el" '(eval-last-sexp :wk "Evaluate elisp expression before point")
    "er" '(eval-region :wk "Evaluate elisp in region"))

  (satori/leader-keys
    "h" '(:ignore t :wk "Help")
    "hd" '(:ignore t :wk "Describe")
    "hdf" '(describe-function :wk "Describe function")
    "hdv" '(describe-variable :wk "Describe variable")
    "hdk" '(describe-key :wk "Describe key")
    "hdb" '(describe-key-briefly :wk "Describe key briefly")
    "hi" '(:ignore t :wk "Info")
    "hii" '(info :wk "Info documentation browser")
    "hid" '(Info-directory :wk "Info Directory")
    "hia" '(info-apropos :wk "Find indice")
    "hr" '(+reload-init :wk "Reload emacs config")
    "ht" '(evil-tutor-start :wk "Evil tutor"))

  (satori/leader-keys
    "o" '(:ignore t :wk "Org")
    "oa" '(org-agenda :wk "Org agenda")
    "os" '((lambda () (interactive) (find-file "~/Documents/org-roam/20231110004145-satori.org")) :wk "Satori's Brain")
    "oe" '(org-export-dispatch :wk "Org export dispatch")
    "oi" '(org-toggle-item :wk "Org toggle item")
    "ot" '(org-todo :wk "Org todo")
    "oB" '(org-babel-tangle :wk "Org babel tangle")
    "oT" '(org-todo-list :wk "Org todo list")
    "ol" '(org-insert-link :wk "Org insert link"))

  (satori/leader-keys
    "r" '(:ignore t :wk "Roam")
    "rf" '(org-roam-node-find :wk "Roam find")
    "ri" '(org-roam-node-insert :wk "Roam insert"))

  (satori/leader-keys
    "ob" '(:ignore t :wk "Tables")
    "ob -" '(org-table-insert-hline :wk "Insert hline in table"))

  (satori/leader-keys
    "od" '(:ignore t :wk "Date/deadline")
    "odt" '(org-time-stamp :wk "Org time stamp"))

  (satori/leader-keys
    "l" '(:ignore t :wk "Misc Temp")
    "ls" '(lsp-treemacs-symbols :wk "Toggle treemacs-symbols"))

  (satori/leader-keys
    "t" '(:ignore t :wk "Toggle")
    "tl" '(display-line-numbers-mode :wk "Toggle line numbers")
    "tt" '(visual-line-mode :wk "Toggle truncated lines")
    "t/" '(vterm-toggle :wk "Toggle vterm")
    "te" '(eshell-toggle :wk "Toggle eshell")
    "tm" '(treemacs :wk "Toggle treemacs"))

  (satori/leader-keys
    "w" '(:ignore t :wk "Windows")
    ;; Window splits
    "wd" '(evil-window-delete :wk "Close window")
    "wa" '(evil-window-new :wk "New window")
    "wv" '(evil-window-vsplit :wk "Vertical split window")
    "ws" '(evil-window-split :wk "Horizontal split window")
    ;; Window motions
    "ww" '(evil-window-next :wk "Goto next window")
    ;; Move Windows
    "wH" '(buf-move-left :wk "Buffer move left")
    "wJ" '(buf-move-down :wk "Buffer move down")
    "wK" '(buf-move-up :wk "Buffer move up")
    "wL" '(buf-move-right :wk "Buffer move right"))
  (general-define-key
   :keymaps 'global
   "C-l" 'evil-window-right
   "C-h" 'evil-window-left
   "C-j" 'evil-window-up
   "C-k" 'evil-window-down)
)

(use-package git-timemachine
  :after git-timemachine
  :hook (evil-normalize-keymaps . git-timemachine-hook)
  :config
  (evil-define-key 'normal git-timemachine-mode-map (kbd "C-j") 'git-timemachine-show-previous-revision)
  (evil-define-key 'normal git-timemachine-mode-map (kbd "C-k") 'git-timemachine-show-next-revision))

(use-package magit)

(add-hook 'eshell-mode-hook (lambda ()
                           (setq display-line-numbers-type nil)
                           (setq mode-line-format nil)))

(use-package eshell-toggle
  :custom
  (eshell-toggle-size-fraction 3)
  (eshell-toggle-use-projectile-root t)
  (eshell-toggle-run-command nil)
  (eshell-toggle-init-function #'eshell-toggle-init-eshell))

(use-package eshell-syntax-highlighting
  :after esh-mode
  :config
  (eshell-syntax-highlighting-global-mode +1))

;; eshell-syntax-highlighting -- adds fish/zsh-like syntax highlighting.
;; eshell-rc-script -- your profile for eshell; like a bashrc for eshell.
;; eshell-aliases-file -- sets an aliases file for the eshell.

(setq eshell-rc-script (concat user-emacs-directory "eshell/profile")
      eshell-aliases-file (concat user-emacs-directory "eshell/aliases")
      eshell-history-size 5000
      eshell-buffer-maximum-lines 5000
      eshell-hist-ignoredups t
      eshell-scroll-to-bottom-on-input t
      eshell-destroy-buffer-when-process-dies t
      eshell-visual-commands'("bash" "fish" "htop" "ssh" "top" "zsh"))

(defun disable-line-numbers ()
    (display-line-numbers-mode -1))

(defun disable-mode-line ()
   (setq mode-line-format nil))

(use-package vterm
  :hook (vterm-mode . disable-line-numbers)
  :hook (vterm-mode . disable-mode-line)
  :config
  (setq shell-file-name "/bin/zsh"
	vterm-max-scrollback 5000))

(use-package vterm-toggle
  :after vterm
  :config
  (setq vterm-toggle-fullscreen-p nil)
  (setq vterm-toggle-scope 'project)
  (add-to-list 'display-buffer-alist
	       '((lambda (buffer-or-name _)
		   (let ((buffer (get-buffer buffer-or-name)))
		     (with-current-buffer buffer
		       (or (equal major-mode 'vterm-mode)
			   (string-prefix-p vterm-buffer-name (buffer-name buffer))))))
		 (display-buffer-reuse-window display-buffer-at-bottom)
		 (reusable-frames . visible)
		 (window-height . 0.3))))

(electric-pair-mode 1)

(use-package lsp-mode
:ensure t)

(add-hook 'c-mode-hook 'lsp)

(use-package treemacs
    :hook (treemacs-mode . disable-line-numbers)
    :hook (treemacs-mode . disable-mode-line)
    :ensure t
    :defer t
    :config
    (progn
        (setq treemacs-collapse-dirs                   (if treemacs-python-executable 3 0)
          treemacs-deferred-git-apply-delay        0.5
          treemacs-directory-name-transformer      #'identity
          treemacs-display-in-side-window          t
          treemacs-eldoc-display                   'simple
          treemacs-file-event-delay                2000
          treemacs-file-extension-regex            treemacs-last-period-regex-value
          treemacs-file-follow-delay               0.2
          treemacs-file-name-transformer           #'identity
          treemacs-follow-after-init               t
          treemacs-expand-after-init               t
          treemacs-find-workspace-method           'find-for-file-or-pick-first
          treemacs-git-command-pipe                ""
          treemacs-goto-tag-strategy               'refetch-index
          treemacs-header-scroll-indicators        '(nil . "^^^^^^")
          treemacs-hide-dot-git-directory          t
          treemacs-indentation                     2
          treemacs-indentation-string              " "
          treemacs-is-never-other-window           nil
          treemacs-max-git-entries                 5000
          treemacs-missing-project-action          'ask
          treemacs-move-forward-on-expand          nil
          treemacs-no-png-images                   nil
          treemacs-no-delete-other-windows         t
          treemacs-project-follow-cleanup          nil
          treemacs-persist-file                    (expand-file-name ".cache/treemacs-persist" user-emacs-directory)
          treemacs-position                        'left
          treemacs-read-string-input               'from-child-frame
          treemacs-recenter-distance               0.1
          treemacs-recenter-after-file-follow      nil
          treemacs-recenter-after-tag-follow       nil
          treemacs-recenter-after-project-jump     'always
          treemacs-recenter-after-project-expand   'on-distance
          treemacs-litter-directories              '("/node_modules" "/.venv" "/.cask")
          treemacs-project-follow-into-home        nil
          treemacs-show-cursor                     nil
          treemacs-show-hidden-files               t
          treemacs-silent-filewatch                nil
          treemacs-silent-refresh                  nil
          treemacs-sorting                         'alphabetic-asc
          treemacs-select-when-already-in-treemacs 'move-back
          treemacs-space-between-root-nodes        t
          treemacs-tag-follow-cleanup              t
          treemacs-tag-follow-delay                1.5
          treemacs-text-scale                      nil
         ;;treemacs-user-mode-line-format           'none
          treemacs-user-header-line-format         nil
          treemacs-wide-toggle-width               70
          treemacs-width                           35
          treemacs-width-increment                 1
          treemacs-width-is-initially-locked       t
          treemacs-workspace-switch-cleanup        nil)
))

(use-package treemacs-evil
  :after (treemacs evil)
  :ensure t)

(use-package treemacs-projectile
  :after (treemacs projectile)
  :ensure t)

(use-package treemacs-icons-dired
  :hook (dired-mode . treemacs-icons-dired-enable-once)
  :ensure t)

(use-package treemacs-magit
  :after (treemacs magit)
  :ensure t)

(use-package lsp-treemacs
  :after lsp)

(with-eval-after-load 'lsp-treemacs
 (setq lsp-treemacs-symbols-position-params
      '((side . right)
        (slot . 2)
        (width . 0.4))))

(defun my-treemacs-setup ()
 (define-key treemacs-mode-map (kbd "<escape>") #'treemacs-quit))
(add-hook 'treemacs-mode-hook #'my-treemacs-setup)

(use-package projectile
  :config
  (projectile-mode 1))

(use-package dashboard
  :ensure t
  :init
  (setq initial-buffer-choice `dashboard-open)
  (setq dashboard-set-heading-icons t)
  (setq dashboard-set-file-icons t)
  (setq dashboard-banner-logo-title "Satori's Emacs")
  (setq dashboard-center-content t)
  (setq dashboard-startup-banner "/home/satori/.config/emacs/images/satori-scaled.png")
  (setq dashboard-items ' ((recents . 5)
			   (projects . 3)
			   (bookmarks . 3)
			   (agenda . 3)
			   (registers . 3)
			   ))
  :config
  (dashboard-setup-startup-hook))

(use-package diminish)

(use-package flycheck
  :ensure t
  :defer t
  :diminish
  :init (global-flycheck-mode))

(use-package company
  :defer 2
  :diminish
  :custom
  (company-begin-commands '(self-insert-command))
  (company-idle-delay .1)
  (company-minimum-prefix-length 2)
  (company-show-numbers t)
  (company-tooltip-align-annotations 't)
  (global-company-mode t))

(use-package company-box
  :after company
  :diminish
  :hook (company-mode . company-box-mode))

(use-package counsel
  :after ivy
  :diminish
  :config
    (counsel-mode)
    (setq ivy-initial-inputs-alist nil))

(use-package ivy
  :bind
  ;; ivy-resume resumes the last Ivy-based completion.
  (("C-c C-r" . ivy-resume)
   ("C-x B" . ivy-switch-buffer-other-window))
  :diminish
  :custom
  (setq ivy-use-virtual-buffers t)
  (setq ivy-count-format "(%d/%d) ")
  (setq enable-recursive-minibuffers t)
  :config
  (ivy-mode))

(use-package all-the-icons-ivy-rich
  :ensure t
  :init (all-the-icons-ivy-rich-mode 1))

(use-package ivy-rich
  :after ivy
  :ensure t
  :init (ivy-rich-mode 1) ;; this gets us descriptions in M-x.
  :custom
  (ivy-virtual-abbreviate 'full
			  ivy-rich-switch-buffer-align-virtual-buffer t
			  ivy-rich-path-style 'abbrev)
  :config
  (ivy-set-display-transformer 'ivy-switch-buffer
			       'ivy-rich-switch-buffer-transformer))

(use-package format-all
  :commands format-all-mode
  :hook (prog-mode . format-all-mode)
  :config
  (setq-default format-all-formatters '(("C"     (astyle "--mode=c"))
                                        ("Shell" (shfmt "-i" "4" "-ci")))))

(use-package which-key
  :diminish
  :config
  (setq which-key-side-window-location 'bottom
	which-key-sort-order #'which-key-key-order-alpha
	which-key-sort-uppercase-first nil
	which-key-add-column-padding 1
	which-key-max-display-columns nil
	which-key-min-display-lines 6
	which-key-side-window-slot -10
	which-key-side-window-max-height 0.25
	which-key-idle-delay 0.25
	which-key-max-description-length 25
	which-key-allow-imprecise-window-fit nil
	which-key-separator " → " )
   (which-key-mode 1))

(setq backup-directory-alist '(("" . "~/.backup/emacs/")))
